### PR TITLE
Clone Commandlet changes

### PR DESCRIPTION
**Pull request by therealsanmah**

### CODE REVIEW

**Comment by avirupch** [comment]
There are merge conflicts, can you please solve the issue?

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj [hunk 1]
```diff
@@ -118,6 +119,7 @@
     <Compile Include="Properties\AssemblyInfo.cs" />
     <Compile Include="ScenarioTests\BackupPolicyTests.cs" />
     <Compile Include="ScenarioTests\BackupTests.cs" />
+    <Compile Include="ScenarioTests\CloneTests.cs" />
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj]: As decided we are not going to checkin tests to GitHub, so this .cs and corresponding .ps1 files should be removed

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/BackupTests.ps1 [hunk 2]
```diff
@@ -667,3 +667,4 @@ function Test-CreateGetRestoreDeleteBackup_Async
     #Cleanup
     CleanupObjects-BackupScenario $deviceName $dcName $acrName $vdName
 }
+
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/BackupTests.ps1]: Please remove this file from PR

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1 [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+
+Import-Module .\BackupTests.ps1
+
+<#
+.SYNOPSIS
+Gets device name to use for the test
+#>
+function Get-SecondDeviceName ()
+{
+	$selectedResource = Select-AzureStorSimpleResource -ResourceName OneSDK-Resource -RegistrationKey "1975530557201809476:eOqMQdvHon3lGwKVYctxZVnwpZcqi8ZS1uyCLJAl6Wg=:JovQDqP1KyWdh4m3mYkdzQ==#4edfc1cde41104e5"
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1]: Better to get first device and second device from a single call, otherwise an additional REST call is involved

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1 [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+}
+
+<#
+.SYNOPSIS
+Tests create, get, clone, delete of backup.
+#>
+function Test-CloneSingleDeviceDefaultName
+{
+	# Unique object names
+	$dcName = Generate-Name("VolumeContainer")
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1]: All the pre-reqs which are not part of the current test scenario should be encapsulated in a single module.

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1 [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+	} while(($backupToRestore -eq $null) -and ($retryCount -lt 10))
+	Assert-NotNull $backupToRestore
+
+	$backupId = $backupToRestore.InstanceId
+	$snapshotToClone = $backupToRestore.Snapshots[0]
+	$volumes = Get-AzureStorSimpleDeviceVolume -DeviceName $deviceName -VolumeName $vdName
+	$volume = $volumes[0]
+	$targetAcrList = $volume.AcrList
+
+	Start-AzureStorSimpleDeviceCloneJob -SourceDeviceName $deviceName -TargetDeviceName $deviceName -BackupId $backupId  -Snapshot $backupToRestore.Snapshots[0] -TargetAccessControlRecords $targetAcrList -Force
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1]: same for cleanup, encapsulate in a reusable module

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1 [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+	} while(($backupToRestore -eq $null) -and ($retryCount -lt 10))
+	Assert-NotNull $backupToRestore
+
+	$backupId = $backupToRestore.InstanceId
+	$snapshotToClone = $backupToRestore.Snapshots[0]
+	$volumes = Get-AzureStorSimpleDeviceVolume -DeviceName $deviceName -VolumeName $vdName
+	$volume = $volumes[0]
+	$targetAcrList = $volume.AcrList
+
+	Start-AzureStorSimpleDeviceCloneJob -SourceDeviceName $deviceName -TargetDeviceName $secondDeviceName -BackupId $backupId  -Snapshot $backupToRestore.Snapshots[0] -TargetAccessControlRecords $targetAcrList -CloneVolumeName $vdName + "_CustomName" -Force
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1]: since the logic for generating CloneVolumeName is being reused, it should be captured in a module instead of hardcoding in every test

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/BackupPolicy/GetAzureStorSimpleDeviceBackupPolicy.cs [hunk 7]
```diff
@@ -32,12 +32,12 @@ public class GetAzureStorSimpleDeviceBackupPolicy:StorSimpleCmdletBase
     {
         private string deviceId = null;
 
-        [Parameter(Position = 0, Mandatory = true, HelpMessage = StorSimpleCmdletHelpMessage.HelpMessageDeviceName)]
+        [Parameter(Position = 0, Mandatory = true, HelpMessage = StorSimpleCmdletHelpMessage.DeviceName)]
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/BackupPolicy/GetAzureStorSimpleDeviceBackupPolicy.cs]: each cmdlet file is getting affected by this change,, can we park this change for a subsequent checkin post v2? otherwise everyone who has forked from AzCiS/dev needs to handle the breaking change in a lot of files, which I don't think would be a good idea.

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+    using Microsoft.WindowsAzure.Management.StorSimple.Models;
+
+    /// <summary>
+    /// Given a backupId, snapshotId and a targetDeviceId , this commandlet will 
+    /// clone it on the given target device.
+    /// </summary>
+    [Cmdlet(VerbsLifecycle.Start, "AzureStorSimpleBackupCloneJob")]
+    public class StartAzureStorSimpleBackupCloneJob : StorSimpleCmdletBase
+    {
+        [Parameter(Mandatory = true, Position = 0, HelpMessage = StorSimpleCmdletHelpMessage.SourceDeviceName)]
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs]: Harsha was mentioning that he needs deviceid to be specifiable in lieu of devicename for cross-device cmdlets, can you please clarify that with him?

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+            }
+        }
+
+        private bool ProcessParameters()
+        {
+            this.sourceDeviceId = StorSimpleClient.GetDeviceId(SourceDeviceName);
+
+            if (this.sourceDeviceId == null)
+            {
+                WriteVerbose(Resources.NotFoundMessageDevice);
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs]: Consider using Resources.NoDeviceFoundWithGivenNameInResourceMessage, which is more user-friendly. You can even remove Resources.NotFoundMessageDevice from Resources file in case no cmdlet is using this.

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+                        {
+                            TargetDeviceId = targetDeviceId,
+                            BackupSetId = BackupId,
+                            SourceSnapshot = Snapshot,
+                            ReturnWorkflowId = true,
+                            TargetVolName = CloneVolumeName,
+                            TargetACRList = TargetAccessControlRecords
+                        };
+                        response = StorSimpleClient.CloneVolume(sourceDeviceId, request);
+                        HandleAsyncTaskResponse(response, "start");
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs]: Consider using HandleDeviceJobResponse to print the cmdlet to query device job status, as available in https://github.com/AzCiS/azure-powershell/pull/4/files. HandleAsyncTaskResponse will print the cmdlet to query async task status instead.

File: src/ServiceManagement/StorSimple/Commands.StorSimple/StorSimpleCmdletHelpMessage.cs [hunk 11]
```diff
@@ -14,70 +14,80 @@
 
 namespace Microsoft.WindowsAzure.Commands.StorSimple
 {
+    using System.Globalization;
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/StorSimpleCmdletHelpMessage.cs]: This change is not advised.

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs (hunk 9 above)
Comment (therealsanmah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs]: Done

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1 (hunk 6 above)
Comment (therealsanmah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1]: Done

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1 (hunk 5 above)
Comment (therealsanmah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1]: Done

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1 (hunk 4 above)
Comment (therealsanmah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1]: Done

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1 (hunk 3 above)
Comment (therealsanmah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/ScenarioTests/CloneTests.ps1]: Done

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj (hunk 1 above)
Comment (therealsanmah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj]: Done

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj [hunk 12]
```diff
@@ -1,5 +1,6 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
 <Project ToolsVersion="12.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\..\..\packages\xunit.runner.visualstudio.2.0.0-rc2-build1043\build\net20\xunit.runner.visualstudio.props" Condition="Exists('..\..\..\packages\xunit.runner.visualstudio.2.0.0-rc2-build1043\build\net20\xunit.runner.visualstudio.props')" />
```
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj]: Have to undo these changes

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/BackupPolicy/GetAzureStorSimpleDeviceBackupPolicy.cs (hunk 7 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/BackupPolicy/GetAzureStorSimpleDeviceBackupPolicy.cs]: I dont think it would be a big change. And there wont be a better time to handle breaking changes. So rather than pushing it for a later release we should do it asap. And i am hoping the people who forked from the dev branch wont make changes to the same the help messages. We should take this change now.

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Commands.StorSimple.csproj [hunk 13]
```diff
@@ -83,12 +83,13 @@
     <Reference Include="Microsoft.WindowsAzure.Management.Scheduler">
       <HintPath>..\..\..\packages\Microsoft.WindowsAzure.Management.Scheduler.6.0.0\lib\net40\Microsoft.WindowsAzure.Management.Scheduler.dll</HintPath>
     </Reference>
-    <Reference Include="Microsoft.WindowsAzure.Management.StorSimple">
```
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Commands.StorSimple.csproj]: Undo this Change.

File: src/ServiceManagement/StorSimple/Commands.StorSimple/StorSimpleCmdletHelpMessage.cs (hunk 11 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/StorSimpleCmdletHelpMessage.cs]: Removed the unused using statement.

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/BackupPolicy/GetAzureStorSimpleDeviceBackupPolicy.cs (hunk 7 above)
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/BackupPolicy/GetAzureStorSimpleDeviceBackupPolicy.cs]: People who have forked their repo from Dev and used these messages in their forked repos would see a breaking hange for this, which I don't think is necessary. If at all necessary, we should make the change after onboarding both onesdk-phase2 and la-migration back onto dev.

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Properties/Resources.resx [hunk 14]
```diff
@@ -438,4 +438,10 @@
   <data name="SuccessMessageSubmitDeviceJob" xml:space="preserve">
     <value>The {0} job is triggered successfully. Please use the command Get-AzureStorSimpleJob -InstanceId {1} for tracking the job's status</value>
   </data>
+  <data name="StartASSBackupCloneJobMessage" xml:space="preserve">
+    <value>Cloning backup with backupId {0}...</value>
+  </data>
+  <data name="StartASSBackupCloneJobWarningMessage" xml:space="preserve">
+    <value>Are you sure you want to clone the backup with backupId{0}?</value>
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Properties/Resources.resx]: space after backupId

File: src/ServiceManagement/StorSimple/Commands.StorSimple/ServiceClients/StorSimpleClient.cs [hunk 15]
```diff
@@ -100,7 +100,7 @@ public void ThrowCloudExceptionDetails(CloudException cloudException)
             throw new InvalidOperationException(
                 string.Format(error.Message,"\n",error.HttpCode,"\n",error.ExtendedCode));
         }
-
+
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/ServiceClients/StorSimpleClient.cs]: you can remove this file from PR

File: src/ServiceManagement/StorSimple/Commands.StorSimple/StorSimpleCmdletHelpMessage.cs [hunk 16]
```diff
@@ -14,79 +14,87 @@
...
+        public const string VolumeId = "The volume identifier.";
+        public const string VolumeMonitoring = "Flag to enable monitoring.";
+        public const string VolumeName = "The volume name.";
+        public const string VolumeOnline = "Is the volume online";
+        public const string VolumeSize = "The size of volume in bytes.";
+        public const string WaitTillComplete = "Wait till the async operation completes.";
+        public const string BackupPolicyName = "Name of the Backup policy that you wish to retrieve. Skip this parameter to get all policies";
+        public const string BackupIdToDelete = "InstanceId of the Backup that needs to be deleted";
+        public const string BackupIdToRestore = "InstanceId of the Backup that needs to be restored";
+        public const string snapshotIdToRestore = "InstanceId of the Snapshot that needs to be restored";
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/StorSimpleCmdletHelpMessage.cs]: SnapshotIdToRestore

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/App.config [hunk 17]
```diff
@@ -8,4 +8,4 @@
     <add key="CisStampId" value="7ef900f6-ad91-497d-a9c5-2742b7ed4a75" />
     <add key="ClientSettingsProvider.ServiceUri" value="" />
   </appSettings>
-</configuration>
\ No newline at end of file
+</configuration>
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/App.config]: changes in test folder can be excluded

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj [hunk 18]
```diff
@@ -13,7 +13,7 @@
     <FileAlignment>512</FileAlignment>
     <SolutionDir Condition="$(SolutionDir) == '' Or $(SolutionDir) == '*Undefined*'">..\..\..\</SolutionDir>
     <RestorePackages>true</RestorePackages>
-    <NuGetPackageImportStamp>8ec1c917</NuGetPackageImportStamp>
+    <NuGetPackageImportStamp>bd7e919e</NuGetPackageImportStamp>
```
Comment (avirupch) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj]: changes in test folder can be excluded

File: src/ServiceManagement/StorSimple/Commands.StorSimple/StorSimpleCmdletHelpMessage.cs (hunk 16 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/StorSimpleCmdletHelpMessage.cs]: Done

File: src/ServiceManagement/StorSimple/Commands.StorSimple/ServiceClients/StorSimpleClient.cs (hunk 15 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/ServiceClients/StorSimpleClient.cs]: This seems to be more like a nit change. Where there is no space between the two methods. I am letting it in

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Properties/Resources.resx (hunk 14 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Properties/Resources.resx]: done

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj (hunk 18 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/Commands.StorSimple.Test.csproj]: Done.

File: src/ServiceManagement/StorSimple/Commands.StorSimple.Test/App.config (hunk 17 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple.Test/App.config]: done!

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/BackupPolicy/GetAzureStorSimpleDeviceBackupPolicy.cs (hunk 7 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/BackupPolicy/GetAzureStorSimpleDeviceBackupPolicy.cs]: I have asked people who are effected by this change, and they are ok with this change. Hence going ahead and making it

File: src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs (hunk 8 above)
Comment (parvezah) [inline src/ServiceManagement/StorSimple/Commands.StorSimple/Cmdlets/Clone/StartAzureStorSimpleBackupCloneJob.cs]: will have to verify with him. As far as i remember, he said it is not required.
